experimental/FTheoryTools/src/AbstractFTheoryModels/attributes.jl

```diff
@@ -1195,33 +1186,222 @@ function chern_class_c2(m::AbstractFTheoryModel; check::Bool = true)
+  # Compute and set h
+  h = c3_t_ambient - cy * chern_class_c2(m; check = check)
+  set_attribute!(m, :chern_class_c3, h)
+  return h
+end
+
+
+
+@doc raw"""
+    chern_class_c4(m::AbstractFTheoryModel; check::Bool = true)
+
+If the elliptically fibered n-fold $Y_n$ underlying the F-theory model in question is given
+as a hypersurface in a toric ambient space, we can compute a cohomology class $h$ on the
+toric ambient space $X_\Sigma$, such that its restriction to $Y_n$ is the 4th Chern class
+$c_4$ of the tangent bundle of $Y_n$. If those assumptions are satisfied, this method returns
+this very cohomology class $h$, otherwise it raises and error.
+
+As of right now, this method is computationally quite expensive for involved toric varieties,
+such as in the example below. Therefore, think carefully if you truly want to compute this quantity.
+
+The computation of the cohomology ring verifies if the toric variety is simplicial and
+complete. The check for it to be complete can be very time consuming. This can be switched
+off by setting the optional argument `check` to the value `false`, as in the example below.
+
+```jldoctest; setup = :(Oscar.LazyArtifacts.ensure_artifact_installed("QSMDB", Oscar.LazyArtifacts.find_artifacts_toml(Oscar.oscardir)))
+julia> qsm_model = literature_model(arxiv_id = "1903.00009", model_parameters = Dict("k" => 4))
+Hypersurface model over a concrete base
+
+julia> h = chern_class_c4(qsm_model; check = false);
+
+julia> is_trivial(h)
+false
+```
+"""
+function chern_class_c4(m::AbstractFTheoryModel; check::Bool = true)
+  @req (m isa WeierstrassModel || m isa GlobalTateModel || m isa HypersurfaceModel) "Fourth Chern class of F-theory model supported for Weierstrass, global Tate and hypersurface models only"
+  @req base_space(m) isa NormalToricVariety "Fourth Chern class of F-theory model currently supported only for toric base"
+  @req ambient_space(m) isa NormalToricVariety "Fourth Chern class of F-theory model currently supported only for toric ambient space"
+
+  # Check if the answer is known
+  if has_attribute(m, :chern_class_c4)
+    return get_attribute(m, :chern_class_c4)
+  end
+
+  # Compute elements of the cohomology ring corresponding to the torus invariant prime divisors
+  q = Oscar._cohomology_ring_with_quotient_map(ambient_space(m); check = false)
+  c_ds = [g.f for g in gens(q[1])]
+  R = parent(c_ds[1])
+  projection_map = q[2]
+  len_c_ds = length(c_ds)
+
+  # Compute the cohomology class corresponding to the hypersurface equation
+  cy = cohomology_class(toric_divisor_class(ambient_space(m), degree(hypersurface_equation(m))))
+
+  # Identify index combination that may be discarded by virtue of the Stanley-Reisner ideal
+  mnf = Oscar._minimal_nonfaces(ambient_space(m))
+  indices = [findall(x -> x == 1, Vector{Int}(mnf[i, :])) for i in 1:Polymake.nrows(mnf)]
+
+  # Compute c4 of the toric ambient space
+  c4_t_ambient = zero(R)
+  for i in 1:len_c_ds - 3
```

Comment:
Just to say: you'll get essentially identical performance if you just write this:
```suggestion
  for i in 1:len_c_ds
```
The advantage is: you avoid  mistakes caused by off-by-one errors and you get more uniform code between c2/c3/c4/...

Reply:
Thank you. Good pointer!